[CWS] CWS-4206 Add DNS response event / NXDOMAIN

kovagsm opened the pull request:
### What does this PR do?

This PR allows the user the possibility to capture an entire DNS packet and create and event based on it, with the entire process and network context. The user can filter based on the DNS response_code, can be especially interesting when wishing to capture a DNS response that returned an NXDOMAIN, but can be used for all the response codes.

In order to be more efficient, a discarder mechanism was added.

### Motivation

### Describe how you validated your changes
* Unit and manual tests

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

YoannGh commented on pkg/security/secl/model/model.go:
```diff
@@ -548,19 +548,29 @@ type ExitEvent struct {
-	Size  uint16 `field:"question.length"`                                                 // SECLDoc[question.length] Definition:`the total DNS request size in bytes`
-	Count uint16 `field:"question.count"`                                                  // SECLDoc[question.count] Definition:`the total count of questions in the DNS request`
+	ID       uint16      `field:"id"`       // SECLDoc[id] Definition:`[Experimental] the DNS request ID`
+	Question DNSQuestion `field:"question"` // SECLDoc[id] Definition:`DNS Question`
```
You don't need to add a SECLDoc comment on `question` fields here because these aren't not terminal fields.

safchain commented on docs/cloud-workload-security/linux_expressions.md:
```diff
@@ -38,6 +38,7 @@ Triggers are events that correspond to types of activity seen by the system. The
 | `chown` | File | A file’s owner was changed | 7.27 |
 | `connect` | Network | A connect was executed | 7.60 |
 | `dns` | Network | A DNS request was sent | 7.36 |
+| `dns_response` | Network | A DNS response was received | 7.6x |
```
could we use `dns.response` instead of dns_response ? to have a more consistent syntax. In case of a response both dns.question and dns.response will be filled :thinking:

kovagsm committed:
If a rule matches, report DNS messages with:
-Network process context
-Network context

If a rule doesn't match, still send a short version without context

Added a new metric for the discarded dns packets

Make tests use packet injection instead of using a DNS query

safchain commented on pkg/security/secl/model/unmarshallers_linux.go:
```diff
@@ -12,6 +12,7 @@ import (
 	"encoding/binary"
 	"encoding/json"
 	"fmt"
+	"github.com/google/gopacket/layers"
```
nit: move this line around line 23

kovagsm replied:
Addressed

safchain commented on pkg/security/tests/dns_response_test.go:
```diff
@@ -0,0 +1,127 @@
+
+const (
+	OK       = uint8(0)
+	NXDOMAIN = uint8(3)
```
I would expose the error code in SECL using constants, like `dns.reponse.code == NXDOMAIN`

kovagsm replied:
Addressed

kovagsm commented on docs/cloud-workload-security/linux_expressions.md:
```diff
@@ -2891,6 +2892,13 @@ Constants: [DNS qtypes](#dns-qtypes)
 
 
 
+### `dns.response.response_code` {#dns-response-response_code-doc}
```
Addressed

kovagsm committed:
Fix failing job

kovagsm commented:
/merge
